dali/pipeline/executor/queue_policy.h

```diff
@@ -138,26 +138,38 @@ struct UniformQueuePolicy {
     }
   }
 
-  void SignalError() {
-    exec_error_ = true;
+  void NotifyAll() {
     ready_cond_.notify_all();
     free_cond_.notify_all();
   }
 
-  bool IsErrorSignaled() {
-    return exec_error_;
+  void SignalStop() {
+    {
+      std::unique_lock<std::mutex> lock(ready_mutex_);
+      ready_stop_ = true;
+    }
+    for (int i = 0; i < static_cast<int>(OpType::COUNT); ++i) {
+      std::lock_guard<std::mutex> l(stage_work_mutex_[i]);
+      stage_work_stop_[i] = true;
+    }
+    NotifyAll();
   }
 
- protected:
-  bool exec_error_ = false;
+  bool IsStopSignaled() {
+    std::lock_guard<std::mutex> l(ready_mutex_);
+    return ready_stop_;
+  }
 
  private:
   std::queue<int> ready_queue_, free_queue_, in_use_queue_;
   std::mutex ready_mutex_, free_mutex_;
   std::condition_variable ready_cond_, free_cond_;
 
-  std::array<std::queue<int>, static_cast<int>(OpType::COUNT)> stage_work_queue_;
-  std::array<std::mutex, static_cast<int>(OpType::COUNT)> stage_work_mutex_;
+  static const int kOpCount = static_cast<int>(OpType::COUNT);
+  std::array<std::queue<int>, kOpCount> stage_work_queue_;
+  std::array<std::mutex, kOpCount> stage_work_mutex_;
+  std::array<bool, kOpCount> stage_work_stop_ = {{false, false, false, false}};
```

Comment:
I would like to add a comment here explaining the need for all the booleans:

Something like:
```
We use a dedicated stop flag for every mutex & condition_varialbe pair, so when using them in cond_var predicate, we know the changes are propagated properly and we won't miss a notify. 
```